src/coreclr/jit/fgopt.cpp

```diff
@@ -4522,6 +4537,314 @@ bool Compiler::fgReorderBlocks(bool useProfile)
+// fgMoveColdBlocks: Move rarely-run blocks to the end of their respective regions.
+//
+// Notes:
+//    Exception handlers are assumed to be cold, so we won't move blocks within them.
+//
+void Compiler::fgMoveColdBlocks()
+{
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** In fgMoveColdBlocks()\n");
+
+        printf("\nInitial BasicBlocks");
+        fgDispBasicBlocks(verboseTrees);
+        printf("\n");
+    }
+#endif // DEBUG
+
+    // In a single pass, find the end blocks of each region that we can use as insertion points for cold blocks
+    //
+    BasicBlock** const lastColdTryBlocks =
+        (compHndBBtabCount == 0) ? nullptr : new (this, CMK_Generic) BasicBlock* [compHndBBtabCount] {};
+    BasicBlock* const lastMainBB        = fgLastBBInMainFunction();
+    BasicBlock*       lastColdMainBlock = nullptr;
+
+    for (BasicBlock* const block : Blocks(fgFirstBB, lastMainBB))
+    {
+        // If a region ends with a call-finally pair, don't split the pair up.
+        // Instead, set the end of the region to the BBJ_CALLFINALLY block in the pair.
+        // Also, don't consider blocks in handler regions.
+        // (If all of some try region X is enclosed in an exception handler,
+        // lastColdTryBlocks[X] will be null. We will handle this case later.)
+        //
+        if (!block->isBBCallFinallyPairTail() && !block->hasHndIndex())
+        {
+            if (block->hasTryIndex())
+            {
+                lastColdTryBlocks[block->getTryIndex()] = block;
+            }
+            else
+            {
+                lastColdMainBlock = block;
+            }
+        }
+    }
+
+    // Returns true if block was moved to before insertionPoint, false otherwise
+    //
+    auto tryMovingBlock = [this](BasicBlock* const block, BasicBlock* const insertionPoint) -> bool {
+        assert(block != nullptr);
+        assert(!block->IsFirst());
+
+        // We don't save insertion points for handler regions,
+        // so if we're in one, insertionPoint will be null.
+        // This shouldn't apply in any other case.
+        //
+        if (insertionPoint == nullptr)
+        {
+            assert(block->hasHndIndex());
+            return false;
+        }
+
+        // If we're in a handler region inside a try region,
+        // we might have an insertion point saved,
+        // but we don't bother with moving blocks within handler regions
+        //
+        if (block->hasHndIndex())
+        {
+            return false;
+        }
+
+        // At this point, block and insertionPoint should be in the same try region,
+        // and should not be in a handler region
+        //
+        assert(BasicBlock::sameEHRegion(block, insertionPoint));
+
+        if (block == insertionPoint)
+        {
+            return false;
+        }
+
+        // Don't move any part of a call-finally pair -- these need to stay together
```

Comment:
Seems like we could handle these, by skipping the pair tail (like you do now) and then moving both if we want to move the pair head?